api-reference/api-introduction/webhooks.md

```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
+
+Step 1: Extract each item from the signature, by splitting on `,` separator.
+
+```js
+items = 't=1655795539264,v1=b6bc2aa82491c32f1cbef0eb52b7ffaa51467ea65a03b5d4ccdcfb9e0941c946'.split(',')
+// [ 't=1655795539264', 'v1=b6bc2aa82491c32f1cbef0eb52b7ffaa51467ea65a03b5d4ccdcfb9e0941c946' ]
+```
+
+Step 2: Extract timestamp from the signature.
+
+```js
+timestamp = 't=1655795539264'.split('=')[1]
+// '1655795539264'
+```
+
+Step 3: Extract HMAC hash from the signature.
+
+```js
+hmac = 'v1=b6bc2aa82491c32f1cbef0eb52b7ffaa51467ea65a03b5d4ccdcfb9e0941c946'.split('=')[1]
+// 'b6bc2aa82491c32f1cbef0eb52b7ffaa51467ea65a03b5d4ccdcfb9e0941c946'
+```
+
+Step 4: Compute the HMAC hash using the webhook secret, signature timestamp and raw request body.
+
+- HAMC key is the webhook secret.
+- HMAC payload is formed by concatenating the timestamp(as numeric string), character `.` and raw request body (encoded as UTF8 string).
+- Use SHA256 algorithm to compute the HMAC hash.
+- HMAC hash is hex encoded.
+
+```js
+computeHmac = Hmac('sha256')
+  .Key(webhookSecret)
+  .Payload(timestamp + '.' + rawRequestBody)
+  .Encode('hex')
+```
+
+Step 5: Compare the computed HMAC hash with the HMAC hash from signature.
+
+```js
+computeHmac === hmac
+```
+
+If the computed HMAC hash matches the HMAC hash in the signature, the webhook request can be considered valid.
+
+## Preventing replay attacks
+
+When an attacker intercepts a webhook request, he can replay the request multiple times with a valid signature.
+
+To mitigate this, Imagekit webhook signature contains a timestamp. The timestamp is generated right before the webhook request is sent to the server.
+
+For retriving this timestamp, the `verify` method in ImagekitSDK returns timestamp & parsed event object.
+
+If timestamp is within the tolerance limit, the request can be considered as valid, else the request must be stalled.
+
+> Optionally, a stronger approach is to use a nonce to prevent replay attacks. You can use Event ID as nonce, it is guaranteed to be unique across all events.
+>
+> You can find the event ID in `id` field of the event object.
 
-Here is the sample code for listing to webhook by ImageKit in Node.js. Use a tool like [Ngrok](https://ngrok.com/) to make your local server accessible from outside for testing webhook implementation. 
+## Sample Codes
 
-```javascript
-app.post('/webhook', express.json({type: 'application/json'}), (request, response) => {
-  const event = request.body;
+{% tabs %}
+{% tab title="ExpressJS" %}
 
-  // Handle the event
+```js
+const express = require("express");
+const { Webhook } = require("imagekit");
+
+// Webhook configs
+const WEBHOOK_ENDPOINT = "/webhook";
+const WEBHOOK_SECRET = "whsec_EvCLlGyctj2WyIMW2r62m3JnyN4SXMsY"; // Copy from Imagekit dashboard
+const WEBHOOK_EXPIRY_DURATION = 60 * 1000; // 60 seconds
+
+// Server configs
+const PORT = 8081;
+
+const app = express();
+
+app.post("/webhook", express.raw({ type: "*/*" }), (req, res) => {
```

Comment:
`{ type: "*/*" }` or `application/json`. We saw that that req.body was still buffer with `application/json` right?